### PR TITLE
Fix issues reported by pyflakes and add pyflakes step to ci, fixes #3756

### LINKED ISSUE #3756: Use pyflakes for static code checks in CI
**Is your feature request related to a problem? Please describe.**

We made great experience in ctapipe and lstchain runnig pyflakes to catch simple static errors before running the unit test suite.

**Describe the solution you'd like**

Run pyflakes on gammapy as part of the CI

**Describe alternatives you've considered**

**Additional context**

Currently, pyflakes reports a large number of issues in gammapy:

```
❯ pyflakes gammapy | wc -l
294
```

However, most of them are related to star imports which make static checking per file for undefined objects impossible.

We should fix these issues to be able to use pyflakes as part of the CI.

### PR DESCRIPTION
**Pull request by maxnoe**
**Description**

This fixes all issues reported by pyflakes in the gammapy package. Most of them were related to star imports and missing `__all__` statements but some also were genuine issues, unused imports, f-strings without placeholders, unused variables and the like.

### CODE REVIEW

**Comment by adonath** [comment]
Thanks @maxnoe! I think the mostly looks fine, except the places where the `*` import is replaced with the explicit imports. While it is in general true that one should avoid the `*` import, it could be considered a false positive in this case. The point is that one does not need both:
- Either the `*` import and `__all__` to define what is actually visible in the sub-package namespace
- Or the explicit imports, but then there is no need for `__all__` anymore, because the explicit imports define what is actually visible in the namespace

I have a slight preference for the old scheme with a combination of `*` and `__all__`, but the explicit import would work equally well, but then one could get rid of the `__all__`?

**Comment by maxnoe** [comment]
>  Either the * import and __all__ to define what is actually visible in the sub-package namespace
>  Or the explicit imports, but then there is no need for __all__ anymore, because the explicit imports define what is actually visible in the namespace

That is only true for python import semantics. It is not true for sphinx automodapi, I think. That one requires `__all__` to correctly specify the classes as it uses that to decide what gets documented, if I am not mistaken. 

I would have to check if the autogenerated docs changed...

**Comment by adonath** [comment]
@maxnoe It seems there is a `autosummary_imported_members` option for `conf.py`, which addresses this? See  `https://github.com/sphinx-doc/sphinx/pull/6212`.

**Comment by maxnoe** [comment]
It seems there is no way of letting pyflakes (or pylint) resolve the star imports. Which is a bit sad...

**Comment by Bultako** [comment]
Thanks @maxnoe 

I think in general this PR is needed, and in particular adding `pyflakes` to the CI, not only for linting purposes but also to build a better API documentation. At this moment some classes are missing from the API docs, and this PR could help in fixing this issue. 

@adonath 
I have tested just adding the `autosummary_imported_members=True` flag to `docs/conf.py` with no better results, still the same classes missing from the API docs. 

I would go with this PR, but it still needs some modifs.
- remove duplicated classes in `__all__` which yields duplication of classes also in the API docs lists.
- sort the classes in `__all__` alphabetically so it could be more easy to find (i.e. see `modelling` API)

**Comment by Bultako** [comment]
other differences found in the API docs produced with this PR:

- `Parameter` and `Parameters` listed in `modeling` but with no links.
- `SourceCatalog3HWC` and `SourceCatalog2HWC ` missing from `catalog` package.
- `TRUNCATION_VALUE` variable missing in `stats`

**Comment by maxnoe** [comment]
> Parameter and Parameters listed in modeling but with no links.

This is probably because they are defined in a higher level module and sphinx is not able to figure that out. That's quite unusual. They are correctly linked above in `gammapy.modeling` but not in `gammapy.modeling.models`.

I think the best solution would be to just remove these two classes from the  lower level namespace `gammapy.modeling.models` and fix the one notebook where they are imported from there.

**Comment by maxnoe** [comment]
> TRUNCATION_VALUE variable missing in stats

Is this really something that should be documented? It's a module level constant in a cython extension.

> SourceCatalog3HWC and SourceCatalog2HWC missing from catalog packag

fixed

**Comment by adonath** [comment]
Thanks @Bultako for checking! So we just probably keep was @maxnoe is proposing here. It's non-ideal because it increases the effort of bookkeeping (e.g. if new classes and methods are added, they have to be added in 2 places in addition to the actual implementation), but complete documentation is more important.

**Comment by bkhelifi** [comment]
@adonath can we precise then this in the Developer documentation? Because it is not so obvious...

File: gammapy/irf/__init__.py [hunk 1]
```diff
@@ -58,3 +58,5 @@
         EDispMap,
     ]
 )
+
+__all__.extend(cls.__name__ for cls in IRF_REGISTRY)
```
Comment (maxnoe) [inline gammapy/irf/__init__.py:62]: Was that intentionally, @Bultako ?

Shouldn't be needed anymore

Comment (Bultako) [reply gammapy/irf/__init__.py:62]: sorry, it was introduced by mistake when resolved conflicts with the master branch. 

you can safely remove it.

**Comment by Bultako** [comment]
Thanks @maxnoe 
There are still some things to fix:
- rebase since it seems there are conflicts with master
- still some work to do in alphabetical order in `__all__` vars (i.e. modeling.models, estimators..) 
- what to do with not linking to Parameter and Parameters? 

From my side if you just rebase we could merge this PR, I could open a new one to fix ordering and the links to Parameters classes.

**Comment by maxnoe** [comment]
>  rebase since it seems there are conflicts with master

Done  

>  still some work to do in alphabetical order in __all__ vars (i.e. modeling.models, estimators..)
  
`modeling.models` was already sorted. `estimators` sorted now.

>what to do with not linking to Parameter and Parameters?

I'd suggest to remove both classes from `gammapy.modeling.models` and fix the one notebook where they are imported from there and not from `gammapy.modeling`

File: gammapy/irf/__init__.py [hunk 2]
```diff
@@ -3,12 +3,45 @@
...
+    "EDispMap",
+    "EffectiveAreaTable2D",
+    "EnergyDependentMultiGaussPSF",
+    "EnergyDispersion2D",
+    "IRF_REGISTRY",
+    "PSF3D",
+    "PSFKernel",
+    "PSFKing",
+    "PSFMap",
+    "ParametricPSF",
```
Comment (Bultako) [inline gammapy/irf/__init__.py:40]: "ParametricPSF" right before "PSF3D"

File: gammapy/modeling/models/spatial.py [hunk 3]
```diff
@@ -22,11 +22,27 @@
...
+    'GeneralizedGaussianSpatialModel',
+    'DiskSpatialModel',
+    'ShellSpatialModel',
+    'Shell2SpatialModel',
+    'ConstantSpatialModel',
+    'ConstantFluxSpatialModel',
+    'TemplateSpatialModel',
+]
+
+
```
Comment (Bultako) [inline gammapy/modeling/models/spatial.py:39]: "ConstantFluxSpatialModel",
     "ConstantSpatialModel",
     "DiskSpatialModel",
     "GaussianSpatialModel",
     "GeneralizedGaussianSpatialModel",
     "PointSpatialModel",
     "SpatialModel",
     "Shell2SpatialModel",
     "ShellSpatialModel",
     "TemplateSpatialModel",

File: gammapy/modeling/models/spectral.py [hunk 4]
```diff
@@ -21,6 +21,33 @@
...
+    'LogParabolaSpectralModel',
+    'LogParabolaNormSpectralModel',
+    'TemplateSpectralModel',
+    'ScaleSpectralModel',
+    'EBLAbsorptionNormSpectralModel',
+    'NaimaSpectralModel',
+    'GaussianSpectralModel',
+]
+
+
```
Comment (Bultako) [inline gammapy/modeling/models/spectral.py:50]: "BrokenPowerLawSpectralModel",
"CompoundSpectralModel",
"ConstantSpectralModel",
"EBLAbsorptionNormSpectralModel",
"ExpCutoffPowerLaw3FGLSpectralModel",
"ExpCutoffPowerLawNormSpectralModel",
"ExpCutoffPowerLawSpectralModel",
"GaussianSpectralModel",
"LogParabolaNormSpectralModel",
"LogParabolaSpectralModel",
"NaimaSpectralModel",
"PiecewiseNormSpectralModel",
"PowerLawNormSpectralModel",
"PowerLawSpectralModel",
"PowerLaw2SpectralModel",
"SmoothBrokenPowerLawSpectralModel",
"ScaleSpectralModel",
"SpectralModel",
"SuperExpCutoffPowerLaw3FGLSpectralModel",
"SuperExpCutoffPowerLaw4FGLSpectralModel",
"TemplateSpectralModel",
"integrate_spectrum",
"scale_plot_flux",

File: gammapy/modeling/models/temporal.py [hunk 5]
```diff
@@ -14,6 +14,18 @@
...
+    'TemporalModel',
+    'ConstantTemporalModel',
+    'LinearTemporalModel',
+    'ExpDecayTemporalModel',
+    'GaussianTemporalModel',
+    'LightCurveTemplateTemporalModel',
+    'PowerLawTemporalModel',
+    'SineTemporalModel',
+]
+
```
Comment (Bultako) [inline gammapy/modeling/models/temporal.py:27]: "ConstantTemporalModel",
"ExpDecayTemporalModel",
"GaussianTemporalModel",
"LightCurveTemplateTemporalModel",
"LinearTemporalModel",
"PowerLawTemporalModel",
"SineTemporalModel",
"TemporalModel",

**Review by Bultako — CHANGES_REQUESTED** [review]
Thanks @maxnoe 
I've written some comments in between the lines of code to finish with the alphabetical sorting. As for your proposed solution to link to the Parameter and Parameters docstrings, you can do it in this PR or in another different one. It's up to you.

File: gammapy/astro/source/__init__.py [hunk 6]
```diff
@@ -1,5 +1,13 @@
...
-from .pulsar import *
-from .pwn import *
-from .snr import *
+from .pulsar import Pulsar, SimplePulsar
+from .pwn import PWN
+from .snr import SNR, SNRTrueloveMcKee
+
+__all__ = [
+    "PWN",
+    "Pulsar",
```
[the remark this answers is not in the document]
Comment (maxnoe) [reply gammapy/astro/source/__init__.py:9]: Aaaah. It's not sorting case insensitive. Will go over it again.

[the remark this answers is not in the document]
Comment (maxnoe) [reply gammapy/astro/source/__init__.py:9]: `%sort iu` for case insensitive and unique in vim

**Comment by maxnoe** [comment]
@Bultako Went over it again, hope it's ok now

**Comment by Bultako** [comment]
@maxnoe 
If we compare the content of the unresolved comments and actual code in the [files changed tab](https://github.com/gammapy/gammapy/pull/3760/files) we can see that the ordering has not been done properly.
Yes, I did it manually.

**Comment by maxnoe** [comment]
> Yes, I did it manually.

Manually according to which rules? You asked for alphabetic sorting. Yet you diverged from it apparently. So we need to define a sorting criteration which developers can follow, if we don't want that to be alphabetic.

File: gammapy/modeling/models/__init__.py [hunk 7]
```diff
@@ -1,13 +1,132 @@
...
+    'SuperExpCutoffPowerLaw4FGLSpectralModel',
+    'TemplateSpatialModel',
+    'TemplateSpectralModel',
+    'TemporalModel',
+    'create_cosmic_ray_spectral_model',
+    'create_crab_spectral_model',
+    'integrate_spectrum',
+    'scale_plot_flux',
+]
+
```
[the remark this answers is not in the document]
Comment (Bultako) [reply gammapy/modeling/models/__init__.py:123]: @maxnoe 
maybe I'm wrong.. but it does not seem the content in `__all__` is sorted alphabetically.

**Comment by maxnoe** [comment]
Ok, sorry, it seems I missed some files due to the limit on how many tabs neovim can open at the same time....

No everything containing an `__all__` should be sorted correctly.


For future reference, this does this automatically:

```
rg __all__ -l | sort | uniq | xargs -i nvim {} +'g/__all__ = \[$/+1,/\]/-1 sort ui' -c wq 
```

File: gammapy/modeling/models/__init__.py [hunk 8]
```diff
@@ -1,13 +1,126 @@
...
+    'SkyModel',
+    'SmoothBrokenPowerLawSpectralModel',
+    'SpatialModel',
+    'SpectralModel',
+    'SuperExpCutoffPowerLaw3FGLSpectralModel',
+    'SuperExpCutoffPowerLaw4FGLSpectralModel',
+    'TemplateSpatialModel',
+    'TemplateSpectralModel',
+    'TemporalModel',
+]
```
Comment (Bultako) [inline gammapy/modeling/models/__init__.py:122]: please, use double quotes as in proposed by [black](https://github.com/psf/black) formatting style

File: gammapy/modeling/models/spatial.py [hunk 9]
```diff
@@ -22,11 +22,27 @@
 from gammapy.utils.scripts import make_path
 from .core import ModelBase
 
+
+__all__ = [
+    'ConstantFluxSpatialModel',
```
Comment (Bultako) [inline gammapy/modeling/models/spatial.py:27]: please, use double quotes as in proposed by black formatting style

File: gammapy/modeling/models/spectral.py [hunk 10]
```diff
@@ -21,6 +21,33 @@
 from .core import ModelBase
 
 
+__all__ = [
+    'BrokenPowerLawSpectralModel',
```
Comment (Bultako) [inline gammapy/modeling/models/spectral.py:25]: please, use double quotes as in proposed by black formatting style

File: gammapy/modeling/models/temporal.py [hunk 11]
```diff
@@ -14,6 +14,18 @@
 from .core import ModelBase
 
 
+__all__ = [
+    'ConstantTemporalModel',
+    'ExpDecayTemporalModel',
```
Comment (Bultako) [inline gammapy/modeling/models/temporal.py:19]: please, use double quotes as in proposed by black formatting style

**Review by Bultako — CHANGES_REQUESTED** [review]
Thanks @maxnoe 
It seems it worked now :)
If you can replace the single quotes by double quotes in the `__all__` lists, we merge it 👍🏻

**Comment by maxnoe** [comment]
Done

**Comment by Bultako** [comment]
Thanks @maxnoe ! 
merging now

**Review by Bultako — APPROVED** [review]

**Pull request merged**